controllers/nstemplateset/space_roles_test.go

```diff
@@ -72,11 +72,19 @@ func TestEnsureSpaceRoles(t *testing.T) {
 				Exists(). // created
 				HasLabel(toolchainv1alpha1.ProviderLabelKey, toolchainv1alpha1.ProviderLabelValue).
 				HasLabel(toolchainv1alpha1.OwnerLabelKey, nsTmplSet.GetName())
-			// verify that the `last-applied-space-roles` annotation was set on namespace
-			lastApplied, err := json.Marshal(nsTmplSet.Spec.SpaceRoles)
-			require.NoError(t, err)
-			AssertThatNamespace(t, "oddity-appstudio", memberClient.Client).
-				HasAnnotation(toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey, string(lastApplied))
+
+			t.Run("update annotation", func(t *testing.T) {
+				// when
+				_, err := mgr.ensure(logger, nsTmplSet)
+
+				// then
+				// verify that the `last-applied-space-roles` annotation was set on namespace
+				require.NoError(t, err)
+				lastApplied, err := json.Marshal(nsTmplSet.Spec.SpaceRoles)
+				require.NoError(t, err)
+				AssertThatNamespace(t, "oddity-appstudio", memberClient.Client).
+					HasAnnotation(toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey, string(lastApplied))
+			})
```

Comment:
just out of curiosity - why the original code couldn't work? why do we need to call the `ensure` method one more time?

Reply:
indeed, it's not needed. I guess I wrote this test with the expectation that we would need an extra reconcile loop once all resources have been created/updated/deleted, but that was _before_ we agreed on the new strategy ("recreate missing resources and that's all"). 
Reverted in b92fadf